lib/charms/tls_certificates_interface/v1/tls_certificates.py

```diff
@@ -884,23 +907,51 @@ def _on_relation_changed(self, event) -> None:
         Returns:
             None
         """
-        requirer_relation_data = _load_unit_relation_data(event.relation.data[event.unit])
-        provider_relation_data = _load_unit_relation_data(event.relation.data[self.model.unit])
+        requirer_relation_data = _load_relation_data(event.relation.data[event.unit])
         if not self._relation_data_is_valid(requirer_relation_data):
             logger.warning(
                 f"Relation data did not pass JSON Schema validation: {requirer_relation_data}"
             )
             return
-        provider_csrs = _get_provider_csrs(event.relation.data[self.model.unit])
-        requirer_csrs = _get_requirer_csrs(event.relation.data[event.unit])
-        for certificate_signing_request in requirer_csrs:
+        provider_csrs = [
+            certificate_creation_request["certificate_signing_request"]
+            for certificate_creation_request in self._provider_certificates
+        ]
+        requirer_unit_csrs = [
+            certificate_creation_request["certificate_signing_request"]
+            for certificate_creation_request in self._requirer_csrs(event.unit)
+        ]
+        for certificate_signing_request in requirer_unit_csrs:
             if certificate_signing_request not in provider_csrs:
                 self.on.certificate_creation_request.emit(
                     certificate_signing_request=certificate_signing_request,
                     relation_id=event.relation.id,
                 )
-        for certificate in provider_relation_data.get("certificates", []):
-            if certificate["certificate_signing_request"] not in requirer_csrs:
+        self._revoke_certificates_for_which_no_csr_exists(relation_id=event.relation.id)
+
+    def _revoke_certificates_for_which_no_csr_exists(self, relation_id: int) -> None:
+        """Revokes certificates for which no unit has a CSR.
+
+        Goes through all generated certificates and compare agains the list of CSRS for all units
+        of a given relationship.
+
+        Args:
+            relation_id (int): Relation id
+
+        Returns:
+            None
+        """
+        certificates_relation = self.model.get_relation(
+            relation_name=self.relationship_name, relation_id=relation_id
+        )
+        if not certificates_relation:
+            raise RuntimeError(f"Relation {self.relationship_name} does not exist")
+        list_of_csrs = []
+        for unit in certificates_relation.units:
+            for csr in self._requirer_csrs(unit):
+                list_of_csrs.append(csr["certificate_signing_request"])
```

Comment:
A bit more pythonic?
```suggestion
            list_of_csrs.extend(csr["certificate_signing_request"] for csr in self._requirer_csrs(unit))
```

Reply:
Done